### ISSUE TITLE
Segmentation fault when parse large (5MB+) schema.graphql file

### ISSUE THREAD

**Issue by olragon** [webonyx/graphql-php#229]
I am using graphql-php `dev-master` to parse 5MB+ schema.graphql (also generated by graphql-php).
PHP can't complete this operation and return `Segmentation fault`.

I have try PHP 7.0, PHP 7.2, PHP 5.6, same error happened. How can I debug or resolve this issue? Any help much appreciated.

---

parser.php
```php
$schema = file_get_contents('./schema.graphql');
$doc = GraphQL\Language\Parser::parse($schema);
```

schema.graphql (expired after 20 downloads or 24 hours)
```graphql
https://send.firefox.com/download/cf0161f083/#mgoK6eYlp_n3uShso-TJvw
```

---

```
> php parser.php
```

```
[1]    1209 segmentation fault  php parser.php
```

---

```bash
> strace php parser.php
```

```
lstat("/home/olragon/projects/xxx/libraries/composer/composer/../webonyx/graphql-php/src/Language/AST/TypeDefinitionNode.php", {st_mode=S_IFREG|0664, st_size=343, ...}) = 0
open("/home/olragon/projects/xxx/libraries/composer/webonyx/graphql-php/src/Language/AST/TypeDefinitionNode.php", O_RDONLY) = 7
fstat(7, {st_mode=S_IFREG|0664, st_size=343, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=343, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=343, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=343, ...}) = 0
mmap(NULL, 343, PROT_READ, MAP_SHARED, 7, 0) = 0x7f0f6b099000
munmap(0x7f0f6b099000, 343)             = 0
close(7)                                = 0
lstat("/home/olragon/projects/xxx/libraries/composer/composer/../webonyx/graphql-php/src/Language/AST/TypeSystemDefinitionNode.php", {st_mode=S_IFREG|0664, st_size=270, ...}) = 0
open("/home/olragon/projects/xxx/libraries/composer/webonyx/graphql-php/src/Language/AST/TypeSystemDefinitionNode.php", O_RDONLY) = 7
fstat(7, {st_mode=S_IFREG|0664, st_size=270, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=270, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=270, ...}) = 0
fstat(7, {st_mode=S_IFREG|0664, st_size=270, ...}) = 0
mmap(NULL, 270, PROT_READ, MAP_SHARED, 7, 0) = 0x7f0f6b099000
munmap(0x7f0f6b099000, 270)             = 0
close(7)                                = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4fa00000
madvise(0x7f0f4fa00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4f800000
madvise(0x7f0f4f800000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4f600000
madvise(0x7f0f4f600000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4f400000
madvise(0x7f0f4f400000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4f200000
madvise(0x7f0f4f200000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4f000000
madvise(0x7f0f4f000000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4ee00000
madvise(0x7f0f4ee00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4ec00000
madvise(0x7f0f4ec00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4ea00000
madvise(0x7f0f4ea00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4e800000
madvise(0x7f0f4e800000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4e600000
madvise(0x7f0f4e600000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4e400000
madvise(0x7f0f4e400000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4e200000
madvise(0x7f0f4e200000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4e000000
madvise(0x7f0f4e000000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4de00000
madvise(0x7f0f4de00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4dc00000
madvise(0x7f0f4dc00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4da00000
madvise(0x7f0f4da00000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4d800000
madvise(0x7f0f4d800000, 2097152, MADV_HUGEPAGE) = 0
mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f0f4d600000
madvise(0x7f0f4d600000, 2097152, MADV_HUGEPAGE) = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x7ffe6c861ff8} ---
+++ killed by SIGSEGV +++
[1]    1637 segmentation fault  strace php parser.php

```

**Comment by vladar** [comment]
I did some testing. Could not reproduce on PHP7, but I got the segfault on PHP5.6 too. Looks like PHP is not happy about how we use LinkedList in Lexer. I had plans to get rid of it, but no ETA yet. 

In order to make it work on PHP7, try following:
1. Use `dev-master` version of the library (I just pushed some fixes to serialize/unserialize)
2. Use `noLocation` option of the parser, which will exclude location data generation (useful for debugging only anyway) e.g.:
```php
$result = \GraphQL\Language\Parser::parse($schema, ['noLocation' => true]);
$arr = \GraphQL\Utils\AST::toArray($result);
```

It may help but in general, I doubt that you will achieve anything useful in PHP with a schema of such size. Exported JSON AST of this schema is `~20mb` (and you'll get similar JSON sizes with any implementation, I tried with graphql-js). 

Simply reading such JSON into PHP variable will take ~200 mb of memory (rough estimate) and half a second. That's only `json_decode`, before we even start any GraphQL operations.

Schema of such size makes sense when it is loaded once and shared between many requests. It just doesn't make sense for PHP's share-nothing + execute-and-die model. Of course, it might be possible to workaround some limitations of this PHP model, but in general, the tool doesn't seem right for your case.

Anyways, let me know about your progress as this journey is useful for the lib anyway.

**Comment by olragon** [comment]
Thank you for your help and super useful/valuable in depth suggestions.

Your estimate about memory size, run time is true.

We have some options here:
- re-architecting this legacy app (Drupal 7) to reduce schema size
- implement Drupal 7 in ReactPHP to share schema
- implement GraphQL in another language (NodeJS, Golang, ...)

For now, I will close this issue, will update if we have something new.

Thank you for your great work.

**Comment by vladar** [comment]
Note that `buildSchema` is not the only option. One way how this would be possible in pure PHP:

1. Create a regular schema with [lazy type loading](http://webonyx.github.io/graphql-php/type-system/schema/#lazy-loading-of-types). On schema initialization, you only need `Query` and `Mutation` types.

2. Write separate cached AST file per each type.

3. When a query is executed, in your type loader - read AST file of the corresponding type and construct the type instance yourself in the runtime. Refer to the source code of the `BuildSchema` class on how this can be done.

Technically you can make some of the `BuildSchema` methods public to allow it build individual types in addition to the whole schema (will appreciate a PR).

So in short, using chunks of schema definition files may work for you. You may prefer holding AST data in shared memory instead of files, but this is something worth benchmarking.

**Comment by vladar** [comment]
@spawnia Why not? Parser simply produces an AST. You can easily parse type files separately and then extract specific type AST from the resulting document and construct the definition in the type loader (only when it is needed by the query).

Alternatively, you could try using [PHP Immutable Arrays](https://blog.blackfire.io/php-7-performance-improvements-immutable-arrays.html) (and [one more document](http://blog.jpauli.tech/2016/04/08/hashtables.html#immutable-arrays)) for AST cache. But it requires code-generation to store AST in PHP files.

I guess the best solution is to combine both of those. As for me, immutable arrays is a must if you build the schema from AST. It just makes so much sense to keep this cache in PHP AST memory zone (which is shared by different requests) and do not copy the entire array all the time.

**Comment by spawnia** [comment]
> You can easily parse type files separately and then extract specific type AST from the resulting document

To be honest, that feels like a bit of a hack. We use an extra class for partial parsing: https://github.com/nuwave/lighthouse/blob/master/src/Schema/AST/PartialParser.php If you are up for it, maybe we can implement something like this in the core library?

Thank you so much for the links, enabling OPCache did the trick - i am now able to parse my schema file in one go easily, and response time was cut in half.
